tests/common/helpers/parallel_utils.py

```diff
@@ -1,5 +1,180 @@
+        if status_value == InitialCheckStatus.SANITY_CHECK_FAILED.value:
+            pt_assert(False, "Leader node sanity check failed. Exiting the test.")
+
+        return status_value == expected_status.value
+
+    def set_status(self, new_status: InitialCheckStatus, is_leader: bool) -> None:
+        if new_status == InitialCheckStatus.TESTS_COMPLETED:
+            status_value, acknowledgments = self._read_state()
+            if status_value == InitialCheckStatus.TESTS_COMPLETED.value:
+                self._write_state(InitialCheckStatus.TESTS_COMPLETED, acknowledgments + 1, is_leader)
+            else:
+                self._write_state(InitialCheckStatus.TESTS_COMPLETED, 1, is_leader)
+
+        else:
+            self._write_state(new_status, 0, is_leader)
+
+    def set_initial_status(self) -> None:
+        with open(self.state_file, 'a+') as f:
+            fcntl.flock(f, fcntl.LOCK_EX)
+            try:
+                f.seek(0)
+                content = f.readlines()
+                if not content:
+                    f.write("{},{},{},{}\n".format(
+                        datetime.utcnow(),
+                        InitialCheckStatus.IDLE.value,
+                        0,
+                        ParallelRole.UNKNOWN.value,
+                    ))
+
+                    f.flush()
+            finally:
+                fcntl.flock(f, fcntl.LOCK_UN)
+
+    def acknowledge_status(self, ack_status: InitialCheckStatus, is_leader: bool) -> None:
+        status_value, acknowledgments = self._read_state()
+        if status_value != ack_status.value:
+            raise Exception("Cannot acknowledge status {} when status is {}".format(ack_status, status_value))
+
+        acknowledgments += 1
+        self._write_state(ack_status, acknowledgments, is_leader)
+
+    def wait_and_acknowledge_status(self, expected_status: InitialCheckStatus, is_leader: bool) -> None:
+        if not self.should_wait:
+            logger.info("Skip waiting and acknowledging status")
+            return
+
+        if wait_until(
+            1800,
+            10,
+            0,
+            self.is_expected_status, expected_status,
+        ):
+            self.acknowledge_status(expected_status, is_leader)
+        else:
+            pt_assert(False, "Timed out waiting for status {}".format(expected_status))
+
+    def is_all_acknowledged(self, ack_status: InitialCheckStatus) -> bool:
+        status_value, acknowledgments = self._read_state()
+        return status_value == ack_status.value and acknowledgments >= self.num_followers
+
+    def wait_for_all_acknowledgments(self, ack_status: InitialCheckStatus) -> None:
+        if not self.should_wait:
+            logger.info("Skip waiting for all acknowledgments")
+            return
+
+        status_to_timeout = {
+            InitialCheckStatus.SETUP_COMPLETED: 120,
+            InitialCheckStatus.TESTS_COMPLETED: 6000,
+            InitialCheckStatus.TEARDOWN_COMPLETED: 120,
+        }
```

Comment:
The wait timeout for `InitialCheckStatus.TESTS_COMPLETED` is not easy to determine. The leader will do `initial_check_state.wait_for_all_acknowledgments(InitialCheckStatus.TESTS_COMPLETED)` after it has finished its actual tests and before it starts the initial checks teardown stage. 

The leader wants to wait for all the followers to finish their actual tests beacuse there a recovery process might be triggered during the sanity check teardown stage. However, anything can happen when the actual tests are running for a follower. I put 100 minutes (6000 seconds) here to get things started because I suppose if a follower takes longer than 100 minutes to finish its actual tests after the leader has finished its actual tests, something must be wrong here and we should just do `pytest.fail()`. Please also share some insights on this.

Reply:
Based on the discussion with Jianquan, we can disable the wait here (e.g. make it `-1` for now) so we can wait infinitely to see what problems we will have in practice. This will make it easier for us to debug because the process will keep hanging there forever. When we have a clear picture of what will happen during the wait time, we can add a reasonable wait time here.